mayastor/src/core/poller.rs

```diff
@@ -0,0 +1,149 @@
+use std::{
+    ffi::{c_void, CString},
+    ptr::NonNull,
+    time::Duration,
+};
+
+use spdk_sys::{
+    spdk_poller,
+    spdk_poller_pause,
+    spdk_poller_register,
+    spdk_poller_register_named,
+    spdk_poller_resume,
+    spdk_poller_unregister,
+};
+
+/// structure holding our function and context
+struct PollCtx<'a>(Box<dyn FnMut() -> i32 + 'a>);
+
+/// indirection to avoid raw pointers at upper layers
+#[inline(always)]
+extern "C" fn _cb(ctx: *mut c_void) -> i32 {
+    let poll = unsafe { &mut *(ctx as *mut PollCtx) };
+    (poll.0)()
+}
+
+/// Poller structure that allows us to pause, stop, resume periodic tasks
+pub struct Poller<'a> {
+    inner: NonNull<spdk_poller>,
+    ctx: NonNull<PollCtx<'a>>,
+    stopped: bool,
+}
+
+impl<'a> Poller<'a> {
+    /// stop the given poller and take ownership of the passed in closure or
+    /// function
+    pub fn stop(mut self) {
+        unsafe {
+            spdk_poller_unregister(&mut self.inner.as_ptr());
+            Box::from_raw(self.ctx.as_ptr());
+            self.stopped = true;
+        }
+    }
+
+    /// pause the given poller
+    pub fn pause(&mut self) {
```

Comment:
Shall we disallow calling spdk_poller_pause() in case poller is stopped when 'self.stopped == true' ?
According to the code of spdk_poller_unregister(), it NULL's the poller (see https://github.com/spdk/spdk/blob/3c2190c214514d7b1d2fb2dd9ae8214dcb36c8c5/lib/thread/thread.c#L1144),
 
but 'spdk_poller_pause()' does not check the pointer against being non-NULL and just blindly dereferences it, which will cause NULL pointer dereference:
https://github.com/spdk/spdk/blob/3c2190c214514d7b1d2fb2dd9ae8214dcb36c8c5/lib/thread/thread.c#L1180

The same suggestion applies to resume(), which uses spdk_poller_resume() with the same behaviour.

Reply:
Good thinking, I was looking into that myself as well, but I do not think it is needed. Perhaps I'm overlooking so please correct me if I'm wrong. `stop(mut self)` will consume self so when I have this:
```rust
let p = poller::Builder::new().with_whatever().build();
// stop takes mut self after this call, and I can't use "p" any more
p.stop();
```

Right?

Reply:
Ah, yes. You are right. After stopping the controller it dissolves.

Reply:
I've updated the comments mike, to add the consuming explicitly.